Client/src/Utils/WdkModel.ts

```diff
@@ -421,8 +421,30 @@ export type UserDataset = {
     display: string;
     version: string;
   };
-  updloaded: number;
-}
+  uploaded: number;
+};
+
+export type UserDatasetUpload = {
+  id: string;
+  datasetId?: number;
+  datasetName: string;
+  summary: string;
+  projects: string[];
+  status: string;
```

Comment:
**question:** Is the full set of statuses known? If so, it might be better to use a union type for this. For example `status: 'awaiting-upload' | 'complete' | 'error';`.

Reply:
They are but there's lots of them and they might change, so I think the UI code doesn't need to know them all.

 I tried to not compare them directly and instead use the boolean properties encoded in the type, except I didn't do that consistently for `awaiting-upload`  - changed it now.

Reply:
I didn't fully put together how this property was being used. We would typically use a utility function in place of the additional properties being created above (e.g., 'isOngoing`).

For example:
```typescript
function isOngoing(upload: UserDatasetUpload): boolean {
  return upload.status !== 'success' && upload.status !== 'rejected' && upload.status !== 'error';
}
```

Doing so allows the response object to be kept in-tact. It also removes some confusion about why there is a 'status' property, in addition to other status-like boolean properties.

Reply:
They can sure be functions - they've been functions, but then I realised that not only `AllUploads.tsx` needs to use them, I didn't want to replicate the logic for them. Would you put them in a separate file or something like that?

This is how they are used now:
```
wbazant@wbazant-19-01:~/dev/WDKClient$ find Client -type f | xargs grep 'isOngoing\|isCancellable\|isSuccessful\|isUserError'
Client/src/Utils/WdkModel.ts:  isOngoing: boolean;
Client/src/Utils/WdkModel.ts:  isCancellable: boolean;
Client/src/Utils/WdkModel.ts:  isSuccessful: boolean;
Client/src/Utils/WdkModel.ts:  isUserError: boolean;
Client/src/Service/Mixins/UserUploadService.ts:    isOngoing: ! upload.status.match(/success|rejected|errored/),
Client/src/Service/Mixins/UserUploadService.ts:    isCancellable: !! upload.status.match(/awaiting-upload/),
Client/src/Service/Mixins/UserUploadService.ts:    isSuccessful: !! upload.status.match(/success/),
Client/src/Service/Mixins/UserUploadService.ts:    isUserError: !! upload.status.match(/rejected/)
Client/src/Views/UserDatasets/AllUploads.tsx:    {upload.isCancellable && <button className="btn" onClick={()=>onClickCancel()}>Cancel upload</button> }
Client/src/Views/UserDatasets/AllUploads.tsx:              { upload.isOngoing
Client/src/Views/UserDatasets/AllUploads.tsx:                  : upload.isSuccessful
Client/src/Views/UserDatasets/AllUploads.tsx:                  : upload.isUserError
Client/src/Views/UserDatasets/AllUploads.tsx:  const ongoingUploads = uploads.filter(u=>u.isOngoing);
Client/src/Views/UserDatasets/AllUploads.tsx:  const finishedUploads = uploads.filter(u=>!u.isOngoing);
Client/src/Controllers/UserDatasetListController.tsx:    const numOngoingUploads = uploads != null ? uploads.filter(upload => upload.isOngoing).length : 0;
```

BTW the reason to match statuses as regex is so that the server can change them into something else and so that the code doesn't need to know exactly what they are.

Reply:
I would probably put them in `UserDatasetUtils.jsx`. That file really should be converted to TypeScript, but that can be done as a follow-up.

FYI, you can annotate JavaScript functions with JSDoc, and TypeScript will use that information for type checking:

```javascript
/**
 * @return {boolean}
 */
export function isOngoing(upload) {
  // ...
}
``` 

The issue with regexes is that they are fuzzier than string comparison. It increases the surface area for bugs to be introduced.

Reply:
Actually at the risk of gaining a poor reputation as a stubborn fool, I don't want to change these properties into utility functions, I think the current solution is right.

Not insisting on the same type as one that is being returned is correct (same with `errorLines` being a `string[]` instead of the relatively complex object) and the properties are more useful.

If UserDatasetUpload could be a class instead of a JSON object, I'd make them methods, and not computed properties, but it's not a big problem that they're pre-computed and the current solution guarantees that the logic is in the same place as where the object is being created. Storing in a utility function in an unrelated file is not going to have that effect.

There are also no bugs in the regexes now, because I checked them. Based on how they look, they're also likely to be right for any status I don't know about.

Reply:
The difference is mostly stylistic. Yours being a little more object oriented style, and mine being a little more function programming style.

Reply:
It is also very common in React/Redux apps to use functions to derive values rather than store them as properties of the same object from which they are derived.

Reply:
Today I thought of making UserUploadStatus a class that would wrap around a the status string, and giving it those properties as methods, but I saw that would be even more inconsistent, because everything else in WdkModel.ts is a combination of basic types.

I guess another thing that's different about pre-computing these properties is that elsewhere, the WdkModel types seem to usually follow the payload, but `UserDatasetUpload` is instead whatever seemed most useful. Maybe that won't age well either. Do you mind if the code gets merged with this inconsistency in it?